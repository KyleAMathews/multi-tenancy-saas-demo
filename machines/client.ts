import { v4 as uuidv4 } from "uuid"

export async function createRequest({ doc, mutator, request }) {
  const requests = doc.getArray(`requests`)
  const requestId = uuidv4()
  let resolveFunc, rejectFunc

  const promise = new Promise((resolve, reject) => {
    // Storing the resolve and reject functions for later use.
    resolveFunc = resolve
    rejectFunc = reject
  })
  function observe(event) {
    const inserted = event.changes.delta.find(
      (item) => Object.keys(item)[0] === `insert`
    )
    const state = inserted.insert[0]
    // TODO only do this in case of user-directed timeout or network
    // disconnect errors e.g. normally we're fine just waiting to go online
    // but the app might want to error immediately if we disconnect or are offline.
    // if (state.error) {
    // requests.unobserve(observe)
    // return rejectFunc(state)
    // }
    if (state.done) {
      requests.unobserve(observe)
      return resolveFunc(state)
    }
  }

  requests.observe(observe)
  requests.push([
    {
      mutator,
      value: `requested`,
      id: requestId,
      done: false,
      clientCreate: new Date().toJSON(),
      request,
      response: {},
    },
  ])

  return promise
}
