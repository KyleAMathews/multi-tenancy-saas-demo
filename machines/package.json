{
  "dependencies": {
    "@libsql/client": "^0.3.4",
    "libsql": "^0.1.12",
    "node-sql-parser": "^4.10.0",
    "uuid": "^9.0.1",
    "yjs": "^13.6.7"
  },
  "devDependencies": {
    "vitest": "^0.34.4"
  },
  "scripts": {
    "test": "vitest"
  }
}
