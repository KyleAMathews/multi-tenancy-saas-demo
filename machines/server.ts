import * as Y from "yjs"

const isServerSync = true
export function listen({ doc, serverConfig }) {
  const requests = doc.getArray(`requests`)
  requests.observe(async (event) => {
    const inserted = event.changes.delta.find(
      (item) => Object.keys(item)[0] === `insert`
    )
    const retain = event.changes.delta.find(
      (item) => Object.keys(item)[0] === `retain`
    )
    const itemArray = retain?.retain || 0
    const state = event.target.get(itemArray)
    if (state.done !== true) {
      if (
        serverConfig.mutators.hasOwnProperty(state.mutator) &&
        serverConfig.mutators[state.mutator] instanceof Function
      ) {
        const mutatorFunc = await serverConfig.mutators[state.mutator]({
          state,
          doc,
        })
        doc.transact(() => {
          state.response = mutatorFunc()

          if (typeof state.response?.error !== `undefined`) {
            state.error = true
          } else {
            state.error = false
          }

          state.value = `responded`
          state.done = true
          state.serverResponded = new Date().toJSON()
          requests.delete(itemArray, 1)
          requests.insert(itemArray, [state])
        })
      } else {
        console.log(`bad request`)
        doc.transact(() => {
          state.error = true
          state.response = {
            error: `A mutator by that name does not exist`,
          }
        })
      }
    }
  })
}
