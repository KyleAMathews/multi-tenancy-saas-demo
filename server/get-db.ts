import path from "path"
import Database from "libsql"

export function getDb(baseDir, dbName) {
  const dbPath = path.join(baseDir, `${dbName}.db`)
  const db = new Database(dbPath)

  return db
}
