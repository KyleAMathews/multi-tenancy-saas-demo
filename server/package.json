{
  "name": "server",
  "module": "index.ts",
  "type": "module",
  "devDependencies": {
    "bun-types": "latest",
    "tsx": "^3.13.0",
    "vitest": "^0.34.5"
  },
  "scripts": {
    "dev": "tsx index.ts",
    "build": "cd ../app-admin && npm run build && cp -r dist ../server/",
    "start": "NODE_ENV=production tsx index.ts",
    "test": "vitest"
  },
  "peerDependencies": {
    "typescript": "^5.0.0"
  },
  "dependencies": {
    "@coffeeandfun/google-profanity-words": "^2.1.0",
    "@libsql/client": "^0.3.5",
    "cors": "^2.8.5",
    "express": "^4.18.2",
    "fs-extra": "^11.1.1",
    "libsql": "^0.1.23",
    "node-sql-parser": "^4.11.0",
    "situated": "^0.0.1",
    "ws": "^8.14.2",
    "yjs": "^13.6.8"
  }
}
