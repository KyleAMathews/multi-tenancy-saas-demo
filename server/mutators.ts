import path from "path"
import Parser from "node-sql-parser"
import fs from "fs-extra"
import { createClient } from "@libsql/client"
const parser = new Parser.Parser()
import * as util from "node:util"
import * as child_process from "node:child_process"
import { mapResultSet } from "./map-sqlite-resultset"
import { ProfanityEngine } from "@coffeeandfun/google-profanity-words"
const profanity = new ProfanityEngine()

const execAsync = util.promisify(child_process.exec)

async function setupDb(db) {
  await db.execute(`CREATE TABLE Todo (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  title TEXT NOT NULL,
  completed BOOLEAN NOT NULL CHECK (completed IN (0, 1))
);`)
  await db.execute(
    `INSERT INTO Todo (title, completed) VALUES ('Go to Grocery Store', 0)`
  )
}

export const serverConfig = ({ adminUrl, adminAuthToken }) => {
  // Create the admin db
  const adminDb = createClient({
    url: adminUrl,
    authToken: adminAuthToken,
  })
  return {
    context: {
      adminDb,
    },
    mutators: {
      ping: async function ({ state, doc }) {
        return function () {
          return { ok: true, request: state.request }
        }
      },
      createDb: async function ({ state, doc }) {
        // Validate app/db name
        const regex = /^[a-zA-Z0-9-_]{1,27}$/
        if (!regex.test(state.request.name)) {
          console.log({
            name: state.request.name,
            test: regex.test(state.request.name),
          })
          return function () {
            return {
              error: `App names must not contain spaces or special characters and be less than 28 characters`,
            }
          }
        }

        const dbs = doc.getMap(`dbs`)
        if (dbs.has(state.request.name)) {
          return function () {
            return { error: `DB Already exists` }
          }
        }

        const isProfane = await profanity.hasCurseWords(
          state.request.name.split(`-`).join(` `).split(`_`).join(` `)
        )
        if (isProfane) {
          return function () {
            return { error: `Profane db names are not allowed.` }
          }
        }

        let createOutput
        let urlOutput
        let tokenOutput
        dbs.set(state.request.name, {
          name: state.request.name,
          state: `INITIALIZING`,
          updatedAt: new Date().toJSON(),
        })
        let command = `turso db create --group demo-multi-tenant-saas ${state.request.name}`
        if (state.request.fromDb) {
          command += ` --from-db=${state.request.fromDb}`
        }
        try {
          createOutput = await execAsync(command)
          dbs.set(state.request.name, {
            name: state.request.name,
            state: `CREATED`,
            updatedAt: new Date().toJSON(),
          })
          urlOutput = await execAsync(
            `turso db show ${state.request.name} --url`
          )
          tokenOutput = await execAsync(
            `turso db tokens create ${state.request.name}`
          )
          dbs.set(state.request.name, {
            name: state.request.name,
            state: `CREATING TABLES`,
            updatedAt: new Date().toJSON(),
          })
        } catch (e) {
          console.log(e)
          return function () {
            return { error: e }
          }
        }
        console.log({ createOutput, urlOutput, tokenOutput })

        const url = urlOutput.stdout.trim()
        const authToken = tokenOutput.stdout.trim()
        console.log({ url, authToken })

        const db = createClient({ url, authToken })

        // Don't need to do setup for cloned dbs.
        if (!state.request.fromDb) {
          await setupDb(db)
        }
        // Async work first and then return func w/ any sync changes.
        // Validate db doesn't exist in both yjs and on disk.
        // Then create db and create table.
        //
        // TODO also a test to run queries.
        const updatedAt = new Date().toJSON()
        await adminDb.execute({
          sql: `INSERT INTO dbs values (:url, :authToken, :state, :name, :updatedAt)`,
          args: {
            url,
            authToken,
            state: `READY`,
            name: state.request.name,
            updatedAt,
          },
        })

        const totals = mapResultSet(
          await db.execute(
            `select completed, count(*) as count from Todo group by completed`
          )
        )
        console.log({ totals })

        return function () {
          dbs.set(state.request.name, {
            url,
            authToken,
            state: `READY`,
            name: state.request.name,
            total: totals.map((row) => row.count).reduce((a, b) => a + b, 0),
            completed: totals.find((row) => row.completed === 1)?.count || 0,
            updatedAt,
          })
          return { url, name: state.request.name }
        }
      },
      deleteDb: async function ({ state, doc }) {
        const dbs = doc.getMap(`dbs`)
        if (!dbs.has(state.request.name)) {
          return function () {
            return {
              error: `DB with the name ${state.request.name} doesn't exists`,
            }
          }
        }

        let destroyOutput
        try {
          destroyOutput = await execAsync(
            `turso db destroy ${state.request.name} --yes`
          )
        } catch (e) {
          console.log(e)
          return function () {
            return { error: e }
          }
        }

        await adminDb.execute({
          sql: `DELETE FROM dbs WHERE name=:name`,
          args: {
            name: state.request.name,
          },
        })

        return function () {
          dbs.delete(state.request.name)
          return { name: state.request.name }
        }
      },
      selectDb: async function ({ state, doc }) {
        try {
          const dbInfo = doc.getMap(`dbs`).get(state.request.name)
          const db = createClient({
            url: dbInfo.url,
            authToken: dbInfo.authToken,
          })
          const ast = parser.astify(state.request.sql)
          if (ast.type === `select`) {
            console.log(`query`, state.request.sql)
            const results = mapResultSet(await db.execute(state.request.sql))
            // Async work first and then return func w/ any sync changes.
            return function () {
              return {
                ok: true,
                results,
              }
            }
          } else {
            return function () {
              return { error: `Only select operations are allowed` }
            }
          }
        } catch (e) {
          console.log(e)
          return function () {
            return { error: e }
          }
        }
      },
    },
  }
}
