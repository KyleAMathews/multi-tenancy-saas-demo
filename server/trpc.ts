import { initTRPC, TRPCError } from "@trpc/server"
import { z } from "zod"
import Parser from "node-sql-parser"
import { createClient } from "@libsql/client"
import * as util from "node:util"
import * as child_process from "node:child_process"
import { mapResultSet } from "./map-sqlite-resultset"
import { ProfanityEngine } from "@coffeeandfun/google-profanity-words"
const profanity = new ProfanityEngine()
console.log(Parser)
const parser = new Parser.Parser()

const execAsync = util.promisify(child_process.exec)

/**
 * Initialization of tRPC backend
 * Should be done only once per backend!
 */
const t = initTRPC.create()
/**
 * Export reusable router and procedure helpers
 * that can be used throughout the router
 */
const router = t.router
const publicProcedure = t.procedure

async function setupDb(db) {
  await db.execute(`CREATE TABLE Todo (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  title TEXT NOT NULL,
  completed BOOLEAN NOT NULL CHECK (completed IN (0, 1))
);`)
  await db.execute(
    `INSERT INTO Todo (title, completed) VALUES ('Go to Grocery Store', 0)`
  )
}

export const appRouter = router({
  ping: publicProcedure.mutation(async () => {
    return {
      transact: () => {
        true
      },
    }
  }),
  createDb: publicProcedure
    .input(
      z.object({
        name: z
          .string()
          .regex(/^[a-zA-Z0-9-_]{1,27}$/, {
            message: `App name must not contain spaces or special characters`,
          })
          .max(28, { message: `App name must be 28 characters or less` })
          .min(3),
        fromDb: z.string().min(3).optional(),
      })
    )
    .mutation(async function ({ input, ctx: { doc, adminDb, transact } }) {
      const dbs = doc.getMap(`dbs`)
      if (dbs.has(input.name)) {
        throw new TRPCError({
          code: `CONFLICT`,
          message: `A db by this name already exists`,
        })
      }

      const isProfane = await profanity.hasCurseWords(
        input.name.split(`-`).join(` `).split(`_`).join(` `)
      )
      if (isProfane) {
        throw new TRPCError({
          code: `BAD_REQUEST`,
          message: `Profane db names are now allowed.`,
        })
      }

      let createOutput
      let urlOutput
      let tokenOutput
      dbs.set(input.name, {
        name: input.name,
        state: `INITIALIZING`,
        updatedAt: new Date().toJSON(),
      })
      let command = `turso db create --group demo-multi-tenant-saas ${input.name}`
      if (input.fromDb) {
        command += ` --from-db=${input.fromDb}`
      }
      try {
        createOutput = await execAsync(command)
        dbs.set(input.name, {
          name: input.name,
          state: `CREATED`,
          updatedAt: new Date().toJSON(),
        })
        urlOutput = await execAsync(`turso db show ${input.name} --url`)
        tokenOutput = await execAsync(`turso db tokens create ${input.name}`)
        dbs.set(input.name, {
          name: input.name,
          state: `CREATING TABLES`,
          updatedAt: new Date().toJSON(),
        })
      } catch (e) {
        console.log(e)
        dbs.delete(input.name)
        throw new TRPCError({
          code: `INTERNAL_SERVER_ERROR`,
          message: `Error creating database or tables`,
          cause: e,
        })
      }
      console.log({ createOutput, urlOutput, tokenOutput })

      const url = urlOutput.stdout.trim()
      const authToken = tokenOutput.stdout.trim()
      console.log({ url, authToken })

      const db = createClient({ url, authToken })

      // Don't need to do setup for cloned dbs.
      if (!input.fromDb) {
        await setupDb(db)
      }
      // Async work first and then return func w/ any sync changes.
      // Validate db doesn't exist in both yjs and on disk.
      // Then create db and create table.
      //
      // TODO also a test to run queries.
      const updatedAt = new Date().toJSON()
      await adminDb.execute({
        sql: `INSERT INTO dbs values (:url, :authToken, :state, :name, :updatedAt)`,
        args: {
          url,
          authToken,
          state: `READY`,
          name: input.name,
          updatedAt,
        },
      })

      const totals = mapResultSet(
        await db.execute(
          `select completed, count(*) as count from Todo group by completed`
        )
      )
      console.log({ totals })

      transact(() => {
        dbs.set(input.name, {
          url,
          authToken,
          state: `READY`,
          name: input.name,
          total: totals.map((row) => row.count).reduce((a, b) => a + b, 0),
          completed: totals.find((row) => row.completed === 1)?.count || 0,
          updatedAt,
        })
      })

      return { url, name: input.name }
      // return function () {
      // dbs.set(input.name, {
      // url,
      // authToken,
      // state: `READY`,
      // name: input.name,
      // total: totals.map((row) => row.count).reduce((a, b) => a + b, 0),
      // completed: totals.find((row) => row.completed === 1)?.count || 0,
      // updatedAt,
      // })
      // return { url, name: input.name }
      // }
    }),
  deleteDb: publicProcedure
    .input(z.object({ name: z.string() }))
    .mutation(async function ({ input, ctx: { doc, adminDb, transact } }) {
      const dbs = doc.getMap(`dbs`)
      if (!dbs.has(input.name)) {
        return function () {
          return {
            error: `DB with the name ${input.name} doesn't exists`,
          }
        }
      }

      let destroyOutput
      try {
        destroyOutput = await execAsync(`turso db destroy ${input.name} --yes`)
      } catch (e) {
        console.log(e)
        return function () {
          return { error: e }
        }
      }

      await adminDb.execute({
        sql: `DELETE FROM dbs WHERE name=:name`,
        args: {
          name: input.name,
        },
      })

      transact(() => {
        dbs.delete(input.name)
      })

      return input.name
    }),
  selectDb: publicProcedure
    .input(z.object({ name: z.string(), sql: z.string() }))
    .query(async function ({ input, ctx: { doc } }) {
      try {
        const dbInfo = doc.getMap(`dbs`).get(input.name)
        const db = createClient({
          url: dbInfo.url,
          authToken: dbInfo.authToken,
        })
        const ast = parser.astify(input.sql)
        if (ast.type === `select`) {
          const results = mapResultSet(await db.execute(input.sql))
          // Async work first and then return func w/ any sync changes.
          return {
            ok: true,
            results,
          }
        } else {
          throw new TRPCError({
            code: `BAD_REQUEST`,
            message: `Only select operations are allowed`,
          })
        }
      } catch (e) {
        console.log(e)
        throw new TRPCError({
          code: `INTERNAL_SERVER_ERROR`,
          cause: e,
        })
      }
    }),
})
// userUpdateName: publicProcedure
// .input(z.object({ id: z.string(), name: z.string() }))
// .mutation(async (opts) => {
// const {
// input,
// ctx: { users },
// } = opts
// let user
// let id
// users.forEach((u, i) => {
// if (u.id === input.id) {
// user = u
// id = i
// }
// })
// const newUser = { ...user, name: input.name }
// return {
// mutations: () => {
// users.delete(id, 1)
// users.insert(id, [newUser])
// },
// response: newUser,
// }
// }),
// })

export type AppRouter = typeof appRouter
