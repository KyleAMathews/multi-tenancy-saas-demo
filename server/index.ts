import cors from "cors"
import path from "path"
import express from "express"
import { WebSocketServer } from "ws"
import { fileURLToPath } from "url"
import fs from "fs-extra"
import { setupWSConnection, getYDoc } from "situated"
import listen from "../machines/server"
import mapResultSet from "../map-sqlite-resultset"
console.log({ mapResultSet })
import Parser from "node-sql-parser"
import { serverConfig } from "./mutators"
import { createClient } from "@libsql/client"

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)

const baseDir =
  process.env.BASE_DATA_DIR || path.resolve(process.cwd(), `.cache`)
const file = path.join(baseDir, `db.json`)
const dbsDir = path.join(baseDir, `dbs`)
console.log({ dbsDir, dir: path.dirname(dbsDir) })
fs.ensureDirSync(path.dirname(file))
fs.ensureDirSync(dbsDir)

const app = express()
app.use(express.json())
app.use(
  cors({
    origin: `http://localhost:5174`,
    credentials: true,
  })
)

// Serve static assets.
app.use(`/`, express.static(path.join(__dirname, `../dist`)))

// handle every other route with index.html, which will contain
// a script tag to your application's JavaScript file(s).
app.get(`*`, function (request, response) {
  response.sendFile(path.resolve(__dirname, `../dist/index.html`))
})

app.post(`/invalidate/:dbName`, async (req, res) => {
  const dbName = req.params.dbName

  const doc = getYDoc(`app-doc`)
  const dbs = doc.getMap(`dbs`)

  if (dbs.has(dbName)) {
    const ydocDb = dbs.get(dbName)
    // Query for total + completed and set.
    const db = createClient({ url: ydocDb.url, authToken: ydocDb.authToken })
    const totals = mapResultSet.mapResultSet(
      await db.execute(
        `select completed, count(*) as count from todo group by completed`
      )
    )
    console.log({ ydocDb, totals })
    doc.transact(() => {
      ydocDb.total = totals.map((row) => row.count).reduce((a, b) => a + b, 0)
      ydocDb.completed = totals.find((row) => row.completed === 1)?.count || 0
      ydocDb.updatedAt = new Date().toJSON()
      dbs.set(dbName, ydocDb)
    })
    console.log({ totals })
    res.send(`ok`)
  }
})

const wsServer = new WebSocketServer({ noServer: true })
wsServer.on(`connection`, setupWSConnection)

let port = 3000
if (process.env.NODE_ENV === `production`) {
  port = 4000
}

const server = app.listen(port, () => {
  console.log(`API listening on port ${port}`)
  const doc = getYDoc(`app-doc`)
  console.log(`got doc`)
  listen.listen({ doc, serverConfig: serverConfig({ dbsDir }) })
})

server.on(`upgrade`, (request, socket, head) => {
  wsServer.handleUpgrade(request, socket, head, (socket) => {
    wsServer.emit(`connection`, socket, request)
  })
})
