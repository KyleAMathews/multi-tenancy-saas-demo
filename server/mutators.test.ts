import { beforeAll, afterAll, describe, expect, test } from "vitest"
import * as Y from "yjs"
import fs from "fs"
import path from "path"
import { listen } from "../machines/server"
import { createRequest } from "../machines/client"
import { serverConfig } from "./mutators"
import * as util from "node:util"
import * as child_process from "node:child_process"

const execAsync = util.promisify(child_process.exec)

let tmpDir: string = ``
let doc
let adminDb
beforeAll(async () => {
  const dirName =
    `test-` + Date.now() + `-` + Math.random().toString(36).substring(2, 7)
  tmpDir = path.join(`/tmp`, dirName)

  try {
    // Create the directory synchronously
    fs.mkdirSync(tmpDir)
  } catch (err) {
    console.error(`Failed to create directory:`, err)
  }

  doc = new Y.Doc()

  const adminDbCredentials = {
    url: `libsql://todos-saas-admin-kyleamathews.turso.io`,
    authToken: `eyJhbGciOiJFZERTQSIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE2OTU2Nzg1MjksImlkIjoiM2JjODcxMDgtNWJlZC0xMWVlLTkwOWItNTIxZDZkZmM4M2VmIn0.ARL4cjEOTRD3J-aQET9nOILCgNr9TE6diAV-Rxv-eWWl0P1cmIuuPVxoBFH32pGJKyM03yL3H-Tnq52VXpQwDA`,
  }

  const { context, mutators } = serverConfig({
    adminUrl: adminDbCredentials.url,
    adminAuthToken: adminDbCredentials.authToken,
  })
  console.log({ mutators })

  adminDb = context.adminDb
  // await adminDb.execute(`CREATE TABLE dbs (
  // url TEXT PRIMARY KEY NOT NULL,
  // authToken TEXT NOT NULL,
  // state TEXT NOT NULL,
  // name TEXT NOT NULL,
  // updatedAt TEXT NOT NULL
  // );
  // `)
  listen({
    doc,
    serverConfig: { mutators },
  })
})

// Cleanup
afterAll(async () => {
  let hasDb = true
  try {
    await execAsync(`turso db list | grep foo`)
  } catch (e) {
    console.log(e)
    hasDb = false
  }
  if (hasDb) {
    const destroyOutput = await execAsync(`turso db destroy foo --yes`)
    console.log({ destroyOutput })
  }

  fs.rmSync(tmpDir, { recursive: true })
})

test(`update a robot name`, async () => {
  const doc = new Y.Doc()

  const robots = doc.getMap(`robots`)
  const robotId = `123`
  robots.set(robotId, { id: robotId, name: `boop` })

  // Config that gets new requests & calls right function.
  const serverConfig = {
    mutators: {
      updateRobotName: async function ({ state, doc }) {
        await new Promise((resolve) => setTimeout(resolve, 3))
        // Async work first and then return func w/ any sync changes.
        return function () {
          const robots = doc.getMap(`robots`)
          const robot = robots.get(state.request.id)
          robot.name = state.request.name
          robots.set(state.request.id, robot)
          return { ok: true }
        }
      },
    },
  }

  listen({ doc, serverConfig })

  let newName = `beep`
  const requestObject = await createRequest({
    doc,
    mutator: `updateRobotName`,
    request: { id: robotId, name: newName },
  })

  const { id, clientCreate, serverResponded, ...toSnapshot } = requestObject
  expect(toSnapshot).toMatchSnapshot()
  expect(requestObject.done).toBeTruthy()
  expect(robots.get(requestObject.request.id).name).toEqual(newName)

  newName = `boop`
  const requestObject2 = await createRequest({
    doc,
    mutator: `updateRobotName`,
    request: { id: robotId, name: newName },
  })

  expect(robots.get(requestObject2.request.id).name).toEqual(newName)
})

function makeid(length) {
  let result = ``
  const characters = `abcdefghijklmnopqrstuvwxyz0123456789`
  const charactersLength = characters.length
  let counter = 0
  while (counter < length) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength))
    counter += 1
  }
  return result
}
describe(`serverConfig`, () => {
  const dbName = `foo-${makeid(3)}`
  const cloneDbName = `${dbName}-clone`
  test(`createDb`, async () => {
    const requestObject = await createRequest({
      doc,
      mutator: `createDb`,
      request: { name: dbName },
    })

    console.log({ requestObject })
    expect(requestObject.response.name).toEqual(dbName)

    expect(Object.keys(requestObject.response)).toMatchSnapshot()
    expect(Object.keys(doc.getMap(`dbs`).get(dbName))).toMatchSnapshot()
    expect(doc.getMap(`dbs`).get(dbName).completed).toEqual(0)
    expect(doc.getMap(`dbs`).get(dbName).total).toEqual(1)

    const result = await adminDb.execute({
      sql: `SELECT * from dbs where name=:name`,
      args: { name: dbName },
    })
    expect(result.rows.length).toEqual(1)
    expect(result.rows[0][3]).toEqual(dbName)
  }, 10000)
  test(`clone`, async () => {
    const requestObject = await createRequest({
      doc,
      mutator: `createDb`,
      request: { name: cloneDbName, fromDb: dbName },
    })

    console.log({ requestObject })
    expect(requestObject.response.name).toEqual(cloneDbName)

    expect(Object.keys(requestObject.response)).toMatchSnapshot()
    expect(Object.keys(doc.getMap(`dbs`).get(cloneDbName))).toMatchSnapshot()

    const result = await adminDb.execute({
      sql: `SELECT * from dbs where name=:name`,
      args: { name: dbName },
    })
    expect(result.rows.length).toEqual(1)
    expect(result.rows[0][3]).toEqual(dbName)
  }, 10000)
  test(`deleteDb`, async () => {
    const requestObject = await createRequest({
      doc,
      mutator: `deleteDb`,
      request: { name: dbName },
    })
    const requestObject2 = await createRequest({
      doc,
      mutator: `deleteDb`,
      request: { name: cloneDbName },
    })
    console.log({
      requestObject: requestObject.response,
      requestObject2: requestObject2.response,
    })

    expect(requestObject.response.name).toEqual(dbName)
    expect(Object.keys(requestObject.response)).toMatchSnapshot()
    expect(fs.existsSync(requestObject.response.dbPath)).toBeFalsy()

    const result = await adminDb.execute({
      sql: `SELECT * from dbs where name=:name`,
      args: { name: dbName },
    })
    expect(result.rows.length).toEqual(0)
  })
})
