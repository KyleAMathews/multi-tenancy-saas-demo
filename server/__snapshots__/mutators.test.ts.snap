// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`serverConfig > clone 1`] = `
[
  "url",
  "name",
]
`;

exports[`serverConfig > clone 2`] = `
[
  "url",
  "authToken",
  "state",
  "name",
  "total",
  "completed",
  "updatedAt",
]
`;

exports[`serverConfig > createDb 1`] = `
[
  "url",
  "name",
]
`;

exports[`serverConfig > createDb 2`] = `
[
  "url",
  "authToken",
  "state",
  "name",
  "total",
  "completed",
  "updatedAt",
]
`;

exports[`serverConfig > deleteDb 1`] = `
[
  "name",
]
`;

exports[`serverConfig > deleteDb 2`] = `
[
  "name",
]
`;

exports[`update a robot name 1`] = `
{
  "done": true,
  "error": false,
  "mutator": "updateRobotName",
  "request": {
    "id": "123",
    "name": "beep",
  },
  "response": {
    "ok": true,
  },
  "value": "responded",
}
`;
